Laravel v10

pxpm opened the pull request:
Adds Laravel 10 support in composer and update test matrix.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/679ca4e0-5852-42de-8ba8-c53f36623359) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/679ca4e0-5852-42de-8ba8-c53f36623359) completed: **No new issues**

tabacitu commented on .github/workflows/coverage.yml:
```diff
@@ -24,13 +24,13 @@ jobs:
       - name: Setup PHP
         uses: shivammathur/setup-php@v2
         with:
-          php-version: 8.1
```
Any reason to bump to PHP 8.2? I see PHP 8.1 is the minimum version supported by L10. Wouldn't it be safer to use the minimum version, just in case?

pxpm replied:
This is just the PHP version used to generate the coverge, since I use PHP 8.2 locally, and usually I am the one pushing tests ... I just updated it to be conformant with my local setup. 

We can revert, no issue.

tabacitu commented on .github/workflows/testing.yml:
```diff
@@ -31,6 +31,15 @@ jobs:
+                      php: 7.3
+                    - laravel: 10.*
+                      php: 7.4
+                    - laravel: 10.*
+                      php: 8.0
+                    - laravel: 10.*
+                      phpunit: 8.*
+        
```
I think it's time to NOT exclude the L10 tests, to make sure they are green. Right?

jnoordsij replied:
I think those should actually be excluded; Laravel 10 only works with PHP >=8.1 (and PHPUnit 9 or 10), so running tests on lower PHP versions will immediately fail on the `composer install` operation.

tabacitu commented:
Pedro, L10 is out, let's finish and merge this please 🙏 Let's make sure tests are green, everything in our repos is still working... and officially provide L10 support 💪

tabacitu commented on .github/workflows/testing.yml:
```diff
@@ -31,6 +31,15 @@ jobs:
                       php: 7.4
                     - laravel: 9.*
                       phpunit: 8.*
+                    - laravel: 10.*
+                      php: 7.3
+                    - laravel: 10.*
+                      php: 7.4
+                    - laravel: 10.*
+                      php: 8.0
+                    - laravel: 10.*
+                      phpunit: 8.*
```
```suggestion
                    - laravel: 10.*
                      php: 7.3
                    - laravel: 10.*
                      php: 7.4
```

jnoordsij commented:
Hey there, thanks for all the work on this, looking forward to the full support for Laravel 10!

Running my own applications test suite on Laravel 10 combined with this branch as Backpack version, I encountered the following exception:
```
TypeError: PDO::prepare(): Argument #1 ($query) must be of type string, Illuminate\Database\Query\Expression given in .../vendor/laravel/framework/src/Illuminate/Database/Connection.php:414
```

This originates from this function: 
https://github.com/Laravel-Backpack/CRUD/blob/01e961bf60f65d9bfc21cd668bd95a65e29cea1b/src/app/Models/Traits/HasEnumFields.php#L16-L37

The breaking change in Laravel 10 is described here: https://laravel.com/docs/10.x/upgrade#database-expressions.

Would be great to address this before merging; preferably with adding some sort of additional test for this particular function? Moreover, it might be worth it to check the full upgrade guide for other possible issues, as the test suite here might not be exhaustive.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/5e4b7840-226e-4afd-9de3-0cbcf52fbc1a) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/7634548a-6e06-4951-bccc-a67d86d5d3f1) was created.

pxpm commented:
Hey @jnoordsij thanks for pointing that out. 

It's not easy to add tests to that function since it only works in mysql, and we use in memory sqlite for testing purposes. 

I've fixed the conflict, and searched our codebase for other possible failing scenarios, I didn't found any 🤞 

Tests are all green now, I've just removed L10+dbal 2.5 from matrix as they are not compatible too.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/5e4b7840-226e-4afd-9de3-0cbcf52fbc1a) completed: **3 new issues**

jnoordsij commented on src/app/Models/Traits/HasEnumFields.php:
```diff
@@ -20,9 +20,13 @@ public static function getPossibleEnumValues($field_name)
 
         $instance = new static(); // create an instance of the model to be able to get the table name
         $connectionName = $instance->getConnectionName();
-
+        $connection = DB::connection($connectionName);
```
```suggestion
        $connection = $instance->getConnection();
```
Small suggestion to make things slightly more simple ;)

pxpm replied:
I was not able to merge your suggestions @jnoordsij 

![image](https://user-images.githubusercontent.com/7188159/219399424-7d89577e-21ae-4095-876a-ffcf0b499930.png)

but I agree with it, thanks for pointing out. I've commit it directly.

jnoordsij reviewed (approved):
@pxpm thanks for the fix, tests are passing again at my end too! 😄 
Left a small suggestion for you.

As for the test, I understand the issue with it. You could try to extract just the query execution part into a separate method and test the rest, but that's probably a bit too much of a hassle. Moreover, let's hope most people will soon migrate to using PHP Enums on their Models everywhere (it's somewhere on my slightly too long to-do list as well... 😓).

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/bc592a97-506c-4a23-a71c-6f64c648c118) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/bc592a97-506c-4a23-a71c-6f64c648c118) completed: **3 new issues**

ziming commented on composer.json:
```diff
@@ -43,7 +43,7 @@
     "require-dev": {
         "phpunit/phpunit": "~8.0|~7.0|~9.0",
```
May want to allow phpunit 10 too

https://laravel.com/docs/10.x/upgrade#updating-dependencies

> Optionally, if you wish to use [PHPUnit 10](https://phpunit.de/announcements/phpunit-10.html), you should delete the processUncoveredFiles attribute from the <coverage> section of your application's phpunit.xml configuration file. Then, update the following dependencies in your application's composer.json file:

> nunomaduro/collision to ^7.0
> phpunit/phpunit to ^10.0

pxpm replied:
I saw that, but even allowing PHPUnit 10 may not work for us unless we drop support for other versions right? 

It would be possible to install L9/L10+PHP unit 9, and those would need the `processUncoveredFiles` right ?

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/b2e5913f-965e-4fdc-8545-bd657edec6f9) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/b2e5913f-965e-4fdc-8545-bd657edec6f9) completed: **4 new issues**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/a28233b1-2b0e-4299-9309-a31dda45f6b9) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/a28233b1-2b0e-4299-9309-a31dda45f6b9) completed: **5 new issues, 1 updated code elements**
